futures-core/src/task/__internal/atomic_waker.rs

```diff
@@ -276,16 +319,29 @@ impl AtomicWaker {
     ///
     /// If a waker has not been registered, this returns `None`.
     pub fn take(&self) -> Option<Waker> {
+        let state = self.state.load(Relaxed);
+        if state == EMPTY || state & WAKING != 0 {
+            // One of:
+            // * no waker inside, nothing to wake
+            // * another process is calling wake now
+
+            if self.state.compare_exchange(state, state, Release, Relaxed).is_ok() {
+                return None;
+            }
```

Comment:
the fallthrough on `else` doesn't seem right to me here

Reply:
You can think of this `if` as a speculative optimization. The code should work without `if` block at all, so falling through on `else` is fine.

(Also note that loaded `state` value is not used after `if` block)